arch/xtensa/core/xtensa_mmu.c

```diff
@@ -805,9 +809,80 @@ void arch_mem_unmap(void *addr, size_t size)
 		va += KB(4);
 	}
 
+#if CONFIG_MP_MAX_NUM_CPUS > 1
+	z_xtensa_mmu_tlb_ipi();
+#endif
+
 	k_spin_unlock(&xtensa_mmu_lock, key);
 }
 
+/* This should be implemented in the SoC layer.
+ * This weak version is here to avoid build errors.
+ */
+void __weak z_xtensa_mmu_tlb_ipi(void)
+{
+}
+
+void z_xtensa_mmu_tlb_shootdown(void)
+{
+	unsigned int key;
+
+	/* Need to lock interrupts to prevent any context
+	 * switching until all the page tables are updated.
+	 * Or else we would be switching to another thread
+	 * and running that with incorrect page tables
+	 * which would result in permission issues.
+	 */
+	key = arch_irq_lock();
+
+	/* We don't have information on which page tables have changed,
+	 * so we just invalidate the cache for all L1 page tables.
+	 */
+	sys_cache_data_invd_range((void *)l1_page_table, sizeof(l1_page_table));
+	sys_cache_data_invd_range((void *)l2_page_tables, sizeof(l2_page_tables));
```

Comment:
Does TLB data live in the dcache?  Pretty sure the hardware doesn't populate the cache on TLB fetches (because, of course, the TLB **is** the cache for fetches of translation data).  Or is the idea here that the local CPU might have unflushed data from having changed the page tables locally?  It might, but if it does that's a race anyway; it's about to IPI in the other direction and someone has to lose.

The only reason I ask is because that invalidation is actually fairly expensive, it has to loop over every line in the page tables.

Reply:
This is called from IPI when another CPU changes the page table contents.